lib/fh_cluster.js

```diff
@@ -62,12 +64,163 @@ module.exports = function fhCluster(workerFunc, optionalNumWorkers, optionalBack
 };
 
+/********************************************/
+/* MASTER - code that powers master process */
+/********************************************/
+
+function startMaster(numWorkers, backoffStrategy, backoffResetTimeout) {
+  //Need to store the number of workers
+  g_numWorkers = numWorkers;
+
+  // set up master process event handlers
+  cluster.on('exit', setupOnExitBackOff(backoffStrategy, backoffResetTimeout));
+  cluster.on('exit', setupBoundsTasks);
+  cluster.on('listening', setOnWorkerListeningHandler);
+  process.on('SIGTERM', cleanMasterShutdown('SIGTERM'));
+  process.on('SIGHUP', cleanMasterShutdown('SIGHUP'));
+  process.on('SIGINT', cleanMasterShutdown('SIGINT'));
+
+  // start workers
+  _.times(numWorkers, function() {
+    cluster.fork();
+  });
+}
+
+function setupOnExitBackOff(strategy, backoffResetTimeout) {
+  return function(worker) {
+    if (worker.suicide) {
+      log.debug('Worker #', worker.id, 'commited suicide, won\'t retry');
+      return;
+    }
+    var nextRetry = strategy.next();
+    log.info('Worker #', worker.id, 'exited. Will retry in',
+      nextRetry, 'ms');
+
+    setTimeout(function() {
+      cluster.fork();
+      resetBackoffResetTimeout(strategy, backoffResetTimeout);
+    }, nextRetry);
+  };
+}
+
+function setOnWorkerListeningHandler(worker, address) {
+  var host = address.address || address.addressType === 4 ? '0.0.0.0' : '::';
+  var addr = host + ':' + address.port;
+  log.info('Cluster worker',  worker.id, 'is now listening at', addr);
+
+  //A new process is listening, need to check if any bound tasks need to be assigned.
+  boundTasks.setUpBoundTasks(singleWorkerTasks, cluster);
+}
+
+function setupBoundsTasks(){
+  //If a worker exits, check that all bound workers are on a running worker.
+  boundTasks.setUpBoundTasks(singleWorkerTasks, cluster);
+}
+
+function resetBackoffResetTimeout(backoffStrategy, timeout) {
+  if (timeout) {
+    clearTimeout(timeout);
+  }
+
+  return setTimeout(function() {
+    backoffStrategy.reset();
+  }, 60*60*1000);
+}
+
+function cleanMasterShutdown(type) {
+  return function (cb) {
+    log.debug('exiting master with ' + type + ' [' + process.pid + ']');
+    eachWorker(function(worker) {
+      worker.send('shutdown');
+    });
+    waitForAllDead(function() {
+      log.debug('exited master with ' + type + ' [' + process.pid + ']');
+      if (cb) return cb();
+      process.exit(0);
+    });
+  }
+}
+
+function waitForAllDead(cb) {
+  var someAlive = false;
+  eachWorker(function(worker) {
+    if (!isWorkerDead(worker)) {
+      log.debug('worker ' + worker.id + ' not dead');
+      someAlive = true;
+      setTimeout(function() {
+        waitForAllDead(cb);
+      }, 50);
+    }
+  });
+  if (!someAlive) {
+    log.debug('all workers dead');
+    cb();
+  }
+}
+
+function isWorkerDead(worker ) {
+  return ('isDead' in worker) ? worker.isDead() : worker.state === 'dead';
+}
+
+function eachWorker(callback) {
+  for (var id in cluster.workers) {
```

Comment:
I don't think the recommended if statement is necessary here, is it? I wonder if there's a way to tell sonarqube not to worry about that here so that we don't hear about it every time we change the file.